src/robot/parsing/model/visitor.py

```diff
@@ -14,29 +14,63 @@
 #  limitations under the License.
 
 import ast
+from typing import Any, Callable, Iterator
 
+from ...utils.notset import NOT_SET, NotSet
 from .statements import Node
 
 
 class VisitorFinder:
+    __cls_finder_cache__: "dict[type[Any], Callable[..., Any]|None|NotSet]"
 
-    def _find_visitor(self, cls):
-        if cls is ast.AST:
+    def __new__(cls, *_args: Any, **_kwargs: Any):  # type: ignore[no-untyped-def]
+        if not hasattr(cls, "__cls_finder_cache__"):
+            cls.__cls_finder_cache__ = {}
+        return super().__new__(cls)
+
+    @classmethod
+    def __find_visitor(cls, node_cls: "type[Any]") -> "Callable[..., Any]|None":
+        if node_cls is ast.AST:
             return None
-        method = 'visit_' + cls.__name__
-        if hasattr(self, method):
-            return getattr(self, method)
-        # Forward-compatibility.
-        if method == 'visit_Return' and hasattr(self, 'visit_ReturnSetting'):
-            return getattr(self, 'visit_ReturnSetting')
-        for base in cls.__bases__:
-            visitor = self._find_visitor(base)
-            if visitor:
-                return visitor
+        method_name = "visit_" + node_cls.__name__
+        method = getattr(cls, method_name, None)
+        if callable(method):
+            return method  # type: ignore[no-any-return]
+        if method_name == "visit_Return":
+            method = getattr(cls, "visit_ReturnSetting", None)
+            if callable(method):
+                return method  # type: ignore[no-any-return]
+        for base in node_cls.__bases__:
+            method = cls._find_visitor_cached(base)
+            if method:
+                return method
         return None
 
+    @classmethod
+    def _find_visitor_cached(cls, node_cls: "type[Any]") -> "Callable[..., Any]|None":
+        result = cls.__cls_finder_cache__.get(node_cls, NOT_SET)
+        if result is NOT_SET:
```

Comment:
What benefits does `NOT_SET` have here? Why isn't `node_cls in cache` enough?

Reply:
I noticed an older comment claiming that using `get` with `NOT_SET` would be faster than using `in`. That sounds a bit odd because both require a dictionary lookup but the former also requires one `is` usage. I decided to test it and at least with Python 3.11 on my machine `in` is a bit faster both when the item exists and when it doesn't. The code is also more idiomatic:
```python
>>> from timeit import timeit
>>>
>>> notset = object()
>>> 
>>> def t1(d, k):
...     v = d.get(k, notset)
...     if v is notset:
...        d[k] = 42
...     return d
... 
>>> def t2(d, k):
...     if k not in d:
...        d[k] = 42
...     return d
...
>>> timeit(lambda: t1({}, 1))
0.12673325699870475
>>> timeit(lambda: t2({}, 1))
0.10616556800232502
>>> timeit(lambda: t1({1: 2}, 1))
0.12745981800253503
>>> timeit(lambda: t2({1: 2}, 1))
0.10945483400428202
```

Reply:
In Python, each dictionary operation starts with:

- Generating the hash of the key
- searching for the hash value in the list of keys

and then do_something with the element....

this is not a problem with small dictionaries and with simple classes where the hash can be easily determined (i.e. int, str, float, type). But if you have big dictionaries and then with an own hash algorithm (the __hash__ method was overwritten) then it gets more complicated and then the method with `notset` is faster, but the one with the `defaultdict` a little bit more. `get` is a method and `[]` is an operator. In most variants an operator is faster.

So your simple test above is not really meaningful ;-)

Since in our case we are dealing with relatively small dictionaries and with a hashkey (type) handled by python, there is not much difference to measure.
That's why I adapted the code according to your wishes.

I can provide you with a simple script to measure visitor performance. I also have a more complex script to measure dictionary accesses, if you are interested ;-)

Reply:
```
>>> def t1(d, k):
...     v = d.get(k, notset)  # << dictionary operation
...     if v is notset:
...        v = d[k] = 42  # << dictionary operation
...     return v
... 
>>> def t2(d, k):
...     if k not in d:  # << dictionary operation
...        d[k] = 42  # << dictionary operation
...     return d[k]  # << dictionary operation
```